onnxruntime/core/providers/cuda/rnn/cudnn_rnn_base.cc

```diff
@@ -239,65 +229,80 @@ Status CudnnRnnBase<T>::ComputeInternal(OpKernelContext* ctx) const {
+      const Tensor* B = ctx->Input<Tensor>(RNN_Input_Index::B);
+      ReorganizeWeights(&W, &R, B, w_data, w_desc);
+    }
+
+    // CUDNN_RNN_DATA_LAYOUT_SEQ_MAJOR_UNPACKED works with CUDNN_RNN_PADDED_IO_ENABLED, so that it will auto fill 0 for the shorter sequences
+    CUDNN_RETURN_IF_ERROR(cudnnSetRNNPaddingMode(rnn_descriptors_.rnn_desc, CUDNN_RNN_PADDED_IO_ENABLED));
+
+    size_t workspace_bytes;
+    CUDNN_RETURN_IF_ERROR(cudnnGetRNNWorkspaceSize(CudnnHandle(), rnn_descriptors_.rnn_desc, gsl::narrow_cast<int>(seq_length), x_desc.data(), &workspace_bytes));
+    auto workspace_cuda = GetScratchBuffer<void>(workspace_bytes);
+
+    if (CUDNN_RNN_RELU == rnn_mode_ || CUDNN_RNN_TANH == rnn_mode_ || nullptr == sequence_lens_data) {
+      CUDNN_RETURN_IF_ERROR(cudnnRNNForwardInference(CudnnHandle(),
+                                                     rnn_descriptors_.rnn_desc,
+                                                     gsl::narrow_cast<int>(seq_length),
+                                                     x_desc.data(),
                                                      x_data_input,
                                                      hx_desc,
                                                      hx_data,
                                                      cx_desc,
                                                      cx_data,
                                                      weight_cached_ ? w_desc_cache_ : w_desc,
                                                      weight_cached_ ? w_data_cache_.get() : w_data.get(),
-                                                     y_desc,
+                                                     y_desc.data(),
                                                      y_data,
                                                      y_h_desc,
                                                      y_h_data,
                                                      y_c_desc,
                                                      y_c_data,
-                                                     nullptr, nullptr, nullptr, nullptr,
-                                                     nullptr, nullptr, nullptr, nullptr,
                                                      workspace_cuda.get(),
                                                      workspace_bytes));
-    // Early terminate for this case since Y data is not required, and Y_h is obtained correctly, no need the following code to retrive Y_h from Y data.
-    if (nullptr == Y) {
-      return Status::OK();
+    } else {
+      CudnnDataTensor x_desc;
+      x_desc.Set(CudnnTensor::GetDataType<CudaT>(), seq_length, batch_size, input_size, sequence_lens_data);
+      CudnnDataTensor y_desc;
+      y_desc.Set(CudnnTensor::GetDataType<CudaT>(), seq_length, batch_size, hidden_size_ * num_directions_, sequence_lens_data);
+
+      CUDNN_RETURN_IF_ERROR(cudnnRNNForwardInferenceEx(CudnnHandle(),
```

Comment:
>cudnnRNNForwardInferenceEx [](start = 28, length = 26)

Seems inference is now protected by mutex. An alternative might be to always create descriptors in stack. We may need some perf test to find out which is better.

Reply:
yes, we may need to try creating descriptor locally. if the perf is good, it would be better to implement a stateless Compute. 

---
In reply to: [310228482](https://github.com/microsoft/onnxruntime/pull/1544#discussion_r310228482) [](ancestors = 310228482)

Reply:
I implemented in another way to create descriptors in stack, and compared the performance, it's much slower with current implementation.

Reply:
Can you confirm the slowdown is from creating tensor descriptor on stack? Is ReorganizeWeights is protected by mutex and shared between threads?

---
In reply to: [311792641](https://github.com/microsoft/onnxruntime/pull/1544#discussion_r311792641) [](ancestors = 311792641)